docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+>(); // Type 'jest.Mock<NetworkConfiguration | undefined, [networkClientId: string]>'
+mockGetNetworkConfigurationByNetworkClientId.mockImplementation(
+  (origin, type) => {},
+);
+// Argument of type '(origin: any, type: any) => void' is not assignable to parameter of type '(networkClientId: string) => NetworkConfiguration | undefined'.
+// Target signature provides too few arguments. Expected 2 or more, but got 1.ts(2345)
+```
+
+#### Acceptable usages of `any`
+
+`any` is only acceptable for very specific usages. The following should be considered an exhaustive list of its valid use cases.
+
+##### `any` may be necessary when assigning new properties to or deleting properties from a generic type at runtime
+
+In most type errors involving property access or runtime property assignment, `any` usage can be avoided by substituting with `as unknown as`.
+
+🚫
+
+```typescript
+for (const key of getKnownPropertyNames(this.internalConfig)) {
+  (this as any)[key] = this.internalConfig[key];
+}
+
+delete addressBook[chainId as any];
+// Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ [chainId: `0x${string}`]: { [address: string]: AddressBookEntry; }; }'.
+//  No index signature with a parameter of type 'string' was found on type '{ [chainId: `0x${string}`]: { [address: string]: AddressBookEntry; }; }'.ts(7053)
+```
+
+✅
+
+```typescript
+for (const key of getKnownPropertyNames(this.internalConfig)) {
+  (this as unknown as typeof this.internalConfig)[key] =
+    this.internalConfig[key];
+}
+
+delete addressBook[chainId as unknown as `0x${string}`];
+```
+
+However, when assigning to a generic type, using `as any` is the only solution.
+
+🚫
+
+```typescript
+(state as RateLimitState<RateLimitedApis>).requests[api][origin] = previous + 1;
+// is generic and can only be indexed for reading.ts(2862)
+```
+
+✅
+
+```typescript
+(state as any).requests[api][origin] = previous + 1;
+```
+
+Even in this case, however, `any` usage might be avoidable by using `Object.assign` or spread operator syntax instead of assignment.
+
+```typescript
+Object.assign(state, {
+  requests: {
+    ...state.requests,
+    [api]: { [origin] },
+  },
+});
+
+{
+  ...state,
+  requests: {
+    ...state.requests,
+    [api]: { [origin] },
+  },
+};
+```
+
+##### `any` may be acceptable to use within generic constraints
+
+✅
+
+```typescript
+class BaseController<
+  ...,
+  messenger extends RestrictedControllerMessenger<N, any, any, string, string>
+> ...
+```
+
+- In general, using `any` in this context is not harmful in the same way that it is in other contexts, as the `any` types only are not directly assigned to any specific variable, and only function as constraints.
+- That said, more specific constraints provide better type safety and intellisense, and should be preferred wherever possible.
+
+##### `any` may be acceptable to use to type the error property in a catch block
```

Comment:
Thoughts on removing this guideline? Using `unknown` seems easy enough to me that I see no reason to use `any` instead, given that we have utility functions to help with narrowing errors to make use of error properties like `message`.

Reply:
Applied here https://github.com/MetaMask/contributor-docs/pull/74/commits/5d81b66240fc5f299e943da9afb8bea27a4a51f1